File: Recursion/Leetcode/Medium/Subsets_II/Subsets_II.cpp
// https://leetcode.com/problems/subsets-ii/
#include<bits/stdc++.h>
using namespace std;
void subsets(vector<int> &nums, vector<int> &sub, vector<vector<int>> &subs, int start)
{
    if (start <= nums.size())
    {
        subs.push_back(sub);
        if (start == nums.size())
        {
            return;
        }
    }
    for (int i = start; i < nums.size(); i++)
    {
        if (i != start && nums[i] == nums[i - 1])
        {
            continue;
        }
        sub.push_back(nums[i]);
        subsets(nums, sub, subs, i + 1);
        sub.pop_back();
    }
}

vector<vector<int>> subsetsWithDup(vector<int> &nums)
{
    sort(nums.begin(), nums.end());
    vector<int> sub;
    vector<vector<int>> subs;
    subsets(nums, sub, subs, 0);
    return subs;
}
int main(void) {
    vector<int> nums = {1, 2, 2};
    vector<vector<int>> ans = subsetsWithDup(nums);
    for(auto v: ans) {
        for(int ele: v) {
            cout << ele << " ";
        }
        cout << endl;
    }
    return 0;
}

File: Recursion/Leetcode/Medium/Subsets_II/Subsets_II.java
// https://leetcode.com/problems/subsets-ii/
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Subsets_II {
    static void generate(int []nums, int index, List<Integer> seq, List<List<Integer>> combo) {
        if(index <= nums.length) {
            combo.add(new ArrayList<>(seq));
        }
        for(int i=index; i<nums.length; i++) {
            if(i != index && nums[i] == nums[i-1]) continue;
            seq.add(nums[i]);
            generate(nums, i+1, seq, combo);
            seq.remove(seq.size() - 1);
        }
    }
    public static List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> list = new ArrayList<>();
        List<Integer> seq = new ArrayList<>();
        generate(nums, 0, seq, list);
        return list;
    }
    public static void main(String[] args) {
        int arr[] = {1, 2, 2};
        List<List<Integer>> list = subsetsWithDup(arr);
        System.out.println(list);
    }
}
